why-pandera/index.qmd

```diff
@@ -0,0 +1,240 @@
+::: column
+#### Benefits
+
+-   Supports both data verification and validation, and can run these
+    checks separately.
+-   Supports `polars` data frames to a large extent (see on the right).
+-   Supports large datasets.
+-   Offers schema inference, although not with `polars`.
+-   Pandera is widely used, extensively tested, and has good
+    documentation.
+:::
+
+::: column
+#### Drawbacks
+
+-   Only data frames are accepted as input, so other formats (e.g. CSV)
+    have to be loaded into a data frame first.
+-   While `polars` is supported, the integration is [not yet
+    complete](https://pandera.readthedocs.io/en/stable/index.html#supported-features).
+    E.g., it cannot yet extract metadata from `polars` data frames.
+-   We would need to write custom code to translate our table metadata
+    from JSON to Pandera schemas in Python. For its own schemas, Pandera
+    provides JSON conversion out of the box.
+:::
+:::::
+
+### Great Expectations
+
+[Great
+Expectations](https://docs.greatexpectations.io/docs/core/introduction/)
+is a larger framework for testing and validating data. It also offers a
+range of other functionality, which includes data visualisation, data
+collation from remote sources, and statistical summary generation. It is
+structured around expectations about the data, which are organised into
+expectation suites.
+
+::::: columns
+::: column
+#### Benefits
+
+-   Supports both data verification and validation, and can run these
+    checks separately.
+-   Supports a wide range of data formats, although not `polars`.
+-   Supports large datasets.
+-   Can generate an expectation suite based on data.
+:::
+
+::: column
+#### Drawbacks
+
+-   No support for `polars`.
+-   We would need to write custom code to translate our table metadata
+    from JSON to expectations in Python. For its own expectations
+    suites, Great Expectations provides JSON conversion out of the box.
+-   The API for declaring expectations matches the structure of the Data
+    Package standard less closely than that of the other options.
+-   Significantly larger and more complex to set up than any of the
+    other options.
+:::
+:::::
+
+### Pydantic
+
+[Pydantic](https://docs.pydantic.dev/latest/) is the most popular
+library for matching data against a schema in Python. Its basic use case
+is describing how data should be structured in a Pydantic model and
+checking an object against this model to confirm that they match. Model
+requirements are expressed using type hints and the matching behaviour
+is highly customisable.
+
+::::: columns
+::: column
+#### Benefits
+
+-   Supports data validation.
+:::
+
+::: column
+#### Drawbacks
+
+-   No out-of-the-box support for data verification.
+-   We would need to translate our JSON metadata into Pydantic models.
+-   Pydantic only accepts dictionary-like objects as input, so data
+    files would need to be loaded into Python manually and fed to the
+    Pydantic model row by row.
```

Comment:
That sounds strange?

Reply:
Well, I guess because it's meant to check the structure of individual objects in memory and not tabular data in a file.